Allow disabling the partition stats index independently of column stats

### Problem

`hoodie.metadata.index.partition.stats.enable` was deprecated in #14165 / #14164 so that the partition stats index is controlled solely by the column stats config (`hoodie.metadata.index.column.stats.enable`), for configuration simplicity. A side effect is that the partition stats index can no longer be disabled independently — enabling column stats always builds partition stats too.

Some table layouts need the column stats index *without* partition stats. Tables composed of **externally created files** (e.g. file formats integrated through Apache XTable) benefit from the column stats index for file-level data skipping, but partition stats generation rebuilds a `FileSystemView` over the external base files (`HoodieMetadataWriteUtils.getFilesToFetchColumnStats`) and is not supported for them — the externally-marked file names do not round-trip back to a fileId, so partition stats generation fails:

```
org.apache.hudi.exception.HoodieException: Failed to generate column stats records for metadata table
  at HoodieMetadataWriteUtils.convertMetadataToPartitionStatRecords(...)
Caused by: java.lang.NullPointerException: element cannot be mapped to a null key
  at AbstractTableFileSystemView.buildFileGroups(...)   // groupingBy(HoodieBaseFile::getFileId)
```

Today the only workaround is to disable column stats entirely, which also loses file-level data skipping.

### Proposed solution

Reintroduce `hoodie.metadata.index.partition.stats.enable` as an **advanced** option while preserving the simple default established by #14165:

- Defaults to `true`, and partition stats still require column stats — when unset, behavior is unchanged (partition stats mirror column stats).
- When explicitly set to `false`, partition stats are skipped even though column stats remain enabled.
- Setting it on while column stats is off has no effect.

This lets external table formats keep column-stats data skipping while cleanly turning the partition stats index off.

### Environment

- Hudi version: master (1.3.0-SNAPSHOT)

--- a/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java
+++ b/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java
@@ -2268,7 +2268,9 @@ public class HoodieWriteConfig extends HoodieConfig {
    * @return {@code true} if the partition stats index is enabled, {@code false} otherwise.
    */
   public boolean isPartitionStatsIndexEnabled() {
-    return isMetadataColumnStatsIndexEnabled();
+    // Delegate to the metadata config (which already requires column stats) so the build and delete
+    // paths stay consistent.
+    return isMetadataTableEnabled() && getMetadataConfig().isPartitionStatsIndexEnabled();
   }
 
   /**

--- a/hudi-common/src/main/java/org/apache/hudi/common/config/HoodieMetadataConfig.java
+++ b/hudi-common/src/main/java/org/apache/hudi/common/config/HoodieMetadataConfig.java
@@ -236,6 +236,18 @@ public final class HoodieMetadataConfig extends HoodieConfig {
           + "For the Spark engine, this config defaults to true (enabled), overriding the base default of false. "
           + "For Flink and Java engines, this remains false by default.");
 
+  public static final ConfigProperty<Boolean> ENABLE_METADATA_INDEX_PARTITION_STATS = ConfigProperty
+      .key(METADATA_PREFIX + ".index.partition.stats.enable")
+      .defaultValue(true)
+      .markAdvanced()
+      .sinceVersion("1.0.0")
+      .withDocumentation("Enable aggregating the column stats index to the partition level in the metadata table, "
+          + "used to prune partitions during index lookups. Partition stats requires the column stats index "
+          + "(" + METADATA_PREFIX + ".index.column.stats.enable) to be enabled and has no effect when column stats "
+          + "is disabled. This is an advanced option that should generally be left enabled; disabling it while "
+          + "keeping column stats enabled is recommended only for tables composed of externally created files "
+          + "(e.g. file formats integrated through Apache XTable), where partition stats are not generated.");
+
   public static final ConfigProperty<Integer> METADATA_INDEX_COLUMN_STATS_FILE_GROUP_COUNT = ConfigProperty
       .key(METADATA_PREFIX + ".index.column.stats.file.group.count")
       .defaultValue(2)
@@ -929,7 +941,11 @@ public final class HoodieMetadataConfig extends HoodieConfig {
   }
 
   public boolean isPartitionStatsIndexEnabled() {
-    return getBooleanOrDefault(ENABLE_METADATA_INDEX_COLUMN_STATS);
+    // Partition stats are derived from the column stats index, so they can only be enabled when column
+    // stats is enabled. Within that constraint the partition stats flag is honoured: it defaults to true
+    // (mirroring column stats) but can be explicitly disabled, e.g. for tables of externally created files.
+    return getBooleanOrDefault(ENABLE_METADATA_INDEX_COLUMN_STATS)
+        && getBooleanOrDefault(ENABLE_METADATA_INDEX_PARTITION_STATS);
   }
 
   public int getPartitionStatsIndexFileGroupCount() {
@@ -1095,6 +1111,11 @@ public final class HoodieMetadataConfig extends HoodieConfig {
 
     public Builder withMetadataIndexColumnStats(boolean enable) {
       metadataConfig.setValue(ENABLE_METADATA_INDEX_COLUMN_STATS, String.valueOf(enable));
+      return this;
+    }
+
+    public Builder withMetadataIndexPartitionStats(boolean enable) {
+      metadataConfig.setValue(ENABLE_METADATA_INDEX_PARTITION_STATS, String.valueOf(enable));
       return this;
     }
 
@@ -1453,13 +1474,6 @@ public final class HoodieMetadataConfig extends HoodieConfig {
       }
     }
   }
-
-  /**
-   * The config is now deprecated. Partition stats are configured using the column stats config itself.
-   */
-  @Deprecated
-  public static final String ENABLE_METADATA_INDEX_PARTITION_STATS =
-      METADATA_PREFIX + ".index.partition.stats.enable";
 
   /**
    * @deprecated Use {@link #ENABLE} and its methods.

--- a/hudi-common/src/main/java/org/apache/hudi/common/config/HoodieMetadataConfig.java
+++ b/hudi-common/src/main/java/org/apache/hudi/common/config/HoodieMetadataConfig.java
@@ -240,7 +240,7 @@ public final class HoodieMetadataConfig extends HoodieConfig {
       .key(METADATA_PREFIX + ".index.partition.stats.enable")
       .defaultValue(true)
       .markAdvanced()
-      .sinceVersion("1.0.0")
+      .sinceVersion("1.3.0")
       .withDocumentation("Enable aggregating the column stats index to the partition level in the metadata table, "
           + "used to prune partitions during index lookups. Partition stats requires the column stats index "
           + "(" + METADATA_PREFIX + ".index.column.stats.enable) to be enabled and has no effect when column stats "

--- a/hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPartitionStatsIndex.scala
+++ b/hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPartitionStatsIndex.scala
@@ -149,6 +149,25 @@ class TestPartitionStatsIndex extends PartitionStatsIndexTestBase {
   }
 
   /**
+   * Test case to validate that on a partitioned table, partition stats can be disabled independently while column
+   * stats remains enabled: the column stats index is still built but the partition stats index is not.
+   */
+  @ParameterizedTest
+  @EnumSource(classOf[HoodieTableType])
+  def testColumnStatsEnabledWithPartitionStatsDisabled(tableType: HoodieTableType): Unit = {
+    val hudiOpts = commonOpts ++ Map(
+      DataSourceWriteOptions.TABLE_TYPE.key -> tableType.name(),
+      HoodieMetadataConfig.ENABLE_METADATA_INDEX_PARTITION_STATS.key -> "false")
+    doWriteAndValidateDataAndPartitionStats(hudiOpts, operation = DataSourceWriteOptions.INSERT_OPERATION_OPT_VAL, saveMode = SaveMode.Overwrite, validate = false)
+    doWriteAndValidateDataAndPartitionStats(hudiOpts, operation = DataSourceWriteOptions.UPSERT_OPERATION_OPT_VAL, saveMode = SaveMode.Append, validate = false)
+    // column stats should be built, but partition stats should not be present in the MDT
+    metaClient = HoodieTableMetaClient.reload(metaClient)
+    val metadataPartitions = metaClient.getTableConfig.getMetadataPartitions
+    assertTrue(metadataPartitions.contains(MetadataPartitionType.COLUMN_STATS.getPartitionPath))
+    assertFalse(metadataPartitions.contains(MetadataPartitionType.PARTITION_STATS.getPartitionPath))
+  }
+
+  /**
    * Test case to do a write with updates and rollback the last instant and validate the partition stats index.
    */
   @ParameterizedTest

